pkg/spoke/managedcluster/claim_controller.go

```diff
@@ -0,0 +1,140 @@
+	recorder events.Recorder) factory.Controller {
+	c := &managedClusterClaimController{
+		clusterName:      clusterName,
+		maxClusterClaims: maxClusterClaims,
+		hubClusterClient: hubClusterClient,
+		hubClusterLister: hubManagedClusterInformer.Lister(),
+		claimLister:      claimInformer.Lister(),
+	}
+
+	return factory.New().
+		WithInformers(claimInformer.Informer()).
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, hubManagedClusterInformer.Informer()).
+		WithSync(c.sync).
+		ToController("ManagedClusterClaimController", recorder)
+}
+
+// sync maintains the cluster claims in status of the managed cluster on hub once it joins the hub.
+func (c managedClusterClaimController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	managedCluster, err := c.hubClusterLister.Get(c.clusterName)
+	if err != nil {
+		return fmt.Errorf("unable to get managed cluster with name %q from hub: %w", c.clusterName, err)
+	}
+
+	// current managed cluster does not join the hub yet, do nothing.
+	if !meta.IsStatusConditionTrue(managedCluster.Status.Conditions, clusterv1.ManagedClusterConditionJoined) {
+		syncCtx.Recorder().Eventf("ManagedClusterIsNotAccepted", "Managed cluster %q does not join the hub yet", c.clusterName)
+		return nil
+	}
+
+	return c.exposeClaims(ctx, syncCtx, managedCluster)
+}
+
+// exposeClaims saves cluster claims fetched on managed cluster into status of the
+// managed cluster on hub. Some of the customized claims might not be exposed once
+// the total number of the claims exceeds the value of `cluster-claims-max`.
+func (c managedClusterClaimController) exposeClaims(ctx context.Context, syncCtx factory.SyncContext,
+	managedCluster *clusterv1.ManagedCluster) error {
+	reservedClaims := []clusterv1.ManagedClusterClaim{}
+	customizedClaims := []clusterv1.ManagedClusterClaim{}
+	clusterClaims, err := c.claimLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("unable to list cluster claims: %w", err)
+	}
+
+	reservedClaimNames := sets.NewString(clusterv1alpha1.ReservedClusterClaimNames[:]...)
+	for _, clusterClaim := range clusterClaims {
+		managedClusterClaim := clusterv1.ManagedClusterClaim{
+			Name:  clusterClaim.Name,
+			Value: clusterClaim.Spec.Value,
+		}
+		if reservedClaimNames.Has(clusterClaim.Name) {
+			reservedClaims = append(reservedClaims, managedClusterClaim)
+			continue
+		}
+		customizedClaims = append(customizedClaims, managedClusterClaim)
+	}
+
+	// sort claims by name
+	sort.SliceStable(reservedClaims, func(i, j int) bool {
+		return reservedClaims[i].Name < reservedClaims[j].Name
+	})
+
+	sort.SliceStable(customizedClaims, func(i, j int) bool {
+		return customizedClaims[i].Name < customizedClaims[j].Name
+	})
+
+	// merge and truncated claims
+	claims := append(reservedClaims, customizedClaims...)
+	if total := len(claims); total > c.maxClusterClaims {
```

Comment:
recommendation: express this as the max number of custom claims.  It seems like you should always be able to update reserved names